src/mlpack/methods/ann/gan/gan.hpp

```diff
@@ -95,47 +93,64 @@ class GAN
   //! Move constructor.
   GAN(GAN&&);
 
+  /**
+   * Prepare the network for the given data.
+   * This function won't actually trigger training process.
+   *
+   * @param trainData The real data.
+   */
+  void ResetData(arma::mat trainData);
+
   // Reset function.
   void Reset();
 
   /**
    * Train function.
    *
+   * @param trainData The real data.
    * @return The final objective of the trained model (NaN or Inf on error).
    */
-  template<typename OptimizerType>
-  double Train(OptimizerType& Optimizer);
+  template<typename Policy = PolicyType, typename OptimizerType>
+  typename std::enable_if<std::is_same<Policy, WGANGP>::value, double>::type
+  Train(arma::mat trainData, OptimizerType& Optimizer);
 
   /**
-   * Evaluate function for the Standard GAN and DCGAN.
-   * This function gives the performance of the Standard GAN or DCGAN on the
-   * current input.
+   * Train function for the Standard GAN and DCGAN.
    *
-   * @param parameters The parameters of the network.
-   * @param i Index of the current input.
-   * @param batchSize Variable to store the present number of inputs.
+   * @param trainData The real data.
+   * @param discriminatorOptimizer Optimizer for discriminator network.
+   * @param generatorOptimizer Optimizer for generator network.
+   * @param maxIterations Number of iterations for which to train GAN.
    */
-  template<typename Policy = PolicyType>
+  template<typename Policy = PolicyType,
+           typename DiscOptimizerType,
+           typename GenOptimizerType>
   typename std::enable_if<std::is_same<Policy, StandardGAN>::value ||
-                          std::is_same<Policy, DCGAN>::value, double>::type
-  Evaluate(const arma::mat& parameters,
-           const size_t i,
-           const size_t batchSize);
+                          std::is_same<Policy, DCGAN>::value, void>::type
+  Train(arma::mat trainData,
+        DiscOptimizerType& discriminatorOptimizer,
+        GenOptimizerType& generatorOptimizer,
+        size_t maxIterations);
 
   /**
-   * Evaluate function for the WGAN.
-   * This function gives the performance of the WGAN on the current input.
+   * Train function for WGAN.
    *
-   * @param parameters The parameters of the network.
-   * @param i Index of the current input.
-   * @param batchSize Variable to store the present number of inputs.
+   * @param trainData The real data.
```

Comment:
Do you think we could improve the wording here? Not sure "real" is the best term.

Reply:
Okay, I have currently kept it as `The data points of real distribution.` Let me know your thoughts regarding it.